server/app/filters/ApiKeyUsageFilter.java

```diff
@@ -0,0 +1,74 @@
+package filters;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import auth.ProfileUtils;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import play.api.libs.concurrent.AkkaSchedulerProvider;
+import play.mvc.EssentialAction;
+import play.mvc.EssentialFilter;
+import play.mvc.Http;
+import scala.concurrent.ExecutionContext;
+import services.apikey.ApiKeyService;
+
+/**
+ * This filter looks for requests with paths that begin with /api and record the usage data for the
+ * relevant API key if it has one. The API key update is scheduled to run in a background through to
+ * reduce latency and to ensure issues with recording the usage do not cause API requests to fail.
+ */
+public class ApiKeyUsageFilter extends EssentialFilter {
+
+  private final AkkaSchedulerProvider akkaSchedulerProvider;
+  private final Provider<ApiKeyService> apiKeyServiceProvider;
+  private final Executor exec;
+  private final Provider<ProfileUtils> profileUtilsProvider;
+
+  @Inject
+  public ApiKeyUsageFilter(
+      AkkaSchedulerProvider akkaSchedulerProvider,
+      Provider<ApiKeyService> apiKeyServiceProvider,
+      Executor exec,
+      Provider<ProfileUtils> profileUtilsProvider) {
+    this.akkaSchedulerProvider = checkNotNull(akkaSchedulerProvider);
+    this.apiKeyServiceProvider = checkNotNull(apiKeyServiceProvider);
+    this.exec = checkNotNull(exec);
+    this.profileUtilsProvider = checkNotNull(profileUtilsProvider);
+  }
+
+  @Override
+  public EssentialAction apply(EssentialAction next) {
+    return EssentialAction.of(
+        (Http.RequestHeader request) ->
+            next.apply(request)
+                .map(
+                    result -> {
+                      if (request.path().startsWith("/api")) {
```

Comment:
Not blocking: Nit: I'd suggest switching the path matching to "/api/" to ensure that a newly added and unrelated route doesn't accidentally match this unnecessarily. While I can't think of anything along these lines off the top of my head, it's likely a good idea to be more specific to avoid head-scratching in the future.

Reply:
done